File: p249.java
/* 
 * Solution to Project Euler problem 249
 * By Nayuki Minase
 * 
 * http://nayuki.eigenstate.org/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */


public final class p249 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p249().run());
	}
	
	
	private static final int LIMIT = 5000;
	private static final long MODULUS = 10000000000000000L;
	
	
	public String run() {
		// Use dynamic programming
		boolean[] isPrime = Library.listPrimality(LIMIT * LIMIT / 2);
		long[] numSubsets = new long[LIMIT * LIMIT / 2];  // numSubsets[i] is the number of subsets with sum i, mod 10^16
		numSubsets[0] = 1;
		int maxSum = 0;  // Sum of all primes seen so far
		
		for (int i = 0; i < LIMIT; i++) {
			if (!isPrime[i])
				continue;
			maxSum += i;
			for (int j = maxSum; j >= i; j--) {
				// Optimization of modulo because we know 0 <= numSubsets[j] + numSubsets[j - i] < 2 * MODULUS
				long temp = numSubsets[j] + numSubsets[j - i];
				if (temp < MODULUS)
					numSubsets[j] = temp;
				else
					numSubsets[j] = temp - MODULUS;
			}
		}
		
		long sum = 0;
		for (int i = 0; i < numSubsets.length; i++) {
			if (isPrime[i])
				sum = (sum + numSubsets[i]) % MODULUS;
		}
		return Long.toString(sum);
	}
	
}

File: python/p249.py
# 
# Solution to Project Euler problem 249
# Copyright (c) Project Nayuki. All rights reserved.
# 
# https://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 

import eulerlib


def compute():
	LIMIT = 5000
	MODULUS = 10**16
	
	# Use dynamic programming. count[i] is the number of subsets of primes with the sum of i, modulo MODULUS.
	count = [0] * (LIMIT**2 // 2)
	count[0] = 1
	s = 0  # Sum of all primes seen so far, and thus the highest index among nonzero entries in 'count'
	for p in eulerlib.list_primes(LIMIT):
		for i in reversed(range(s + 1)):
			count[i + p] = (count[i + p] + count[i]) % MODULUS
		s += p
	
	isprime = eulerlib.list_primality(s + 1)
	ans = sum(count[i] for i in range(s + 1) if isprime[i]) % MODULUS
	return str(ans)


if __name__ == "__main__":
	print(compute())
